File: src/ts/binary-writer.ts
/**
 * r[impl jetstream.wireformat.ts.writer]
 * r[impl jetstream.wireformat.byte-order]
 */

/**
 * BinaryWriter writes to a growable buffer.
 * All multi-byte writes use little-endian byte order per 9P2000.L conventions.
 */
export class BinaryWriter {
  private buffer: Uint8Array;
  private view: DataView;
  private offset: number;

  constructor(initialCapacity: number = 256) {
    this.buffer = new Uint8Array(initialCapacity);
    this.view = new DataView(this.buffer.buffer);
    this.offset = 0;
  }

  private ensureCapacity(additionalBytes: number): void {
    const required = this.offset + additionalBytes;
    if (required <= this.buffer.byteLength) return;
    let newCapacity = this.buffer.byteLength;
    while (newCapacity < required) {
      newCapacity *= 2;
    }
    const newBuffer = new Uint8Array(newCapacity);
    newBuffer.set(this.buffer);
    this.buffer = newBuffer;
    this.view = new DataView(this.buffer.buffer);
  }

  writeBytes(bytes: Uint8Array): void {
    this.ensureCapacity(bytes.byteLength);
    this.buffer.set(bytes, this.offset);
    this.offset += bytes.byteLength;
  }

  writeU8(value: number): void {
    this.ensureCapacity(1);
    this.view.setUint8(this.offset, value);
    this.offset += 1;
  }

  writeU16(value: number): void {
    this.ensureCapacity(2);
    this.view.setUint16(this.offset, value, true);
    this.offset += 2;
  }

  writeU32(value: number): void {
    this.ensureCapacity(4);
    this.view.setUint32(this.offset, value, true);
    this.offset += 4;
  }

  writeU64(value: bigint): void {
    this.ensureCapacity(8);
    this.view.setBigUint64(this.offset, value, true);
    this.offset += 8;
  }

  writeI16(value: number): void {
    this.ensureCapacity(2);
    this.view.setInt16(this.offset, value, true);
    this.offset += 2;
  }

  writeI32(value: number): void {
    this.ensureCapacity(4);
    this.view.setInt32(this.offset, value, true);
    this.offset += 4;
  }

  writeI64(value: bigint): void {
    this.ensureCapacity(8);
    this.view.setBigInt64(this.offset, value, true);
    this.offset += 8;
  }

  writeF32(value: number): void {
    this.ensureCapacity(4);
    this.view.setFloat32(this.offset, value, true);
    this.offset += 4;
  }

  writeF64(value: number): void {
    this.ensureCapacity(8);
    this.view.setFloat64(this.offset, value, true);
    this.offset += 8;
  }

  toUint8Array(): Uint8Array {
    return this.buffer.slice(0, this.offset);
  }
}

File: Sources/JetStreamWireFormat/BinaryWriter.swift
// JetStream BinaryWriter
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

import Foundation

// r[impl jetstream.wireformat.swift.writer]
/// A writer that appends binary data to a growable `Data` buffer.
public struct BinaryWriter {
    public private(set) var data: Data

    /// Creates a new BinaryWriter with the given initial capacity.
    public init(capacity: Int = 256) {
        self.data = Data()
        self.data.reserveCapacity(capacity)
    }

    /// Appends raw bytes to the buffer.
    public mutating func writeBytes(_ bytes: Data) {
        data.append(bytes)
    }

    /// Writes a single UInt8.
    public mutating func writeU8(_ value: UInt8) {
        data.append(value)
    }

    // r[impl jetstream.wireformat.byte-order]
    /// Writes a UInt16 in little-endian byte order.
    public mutating func writeU16(_ value: UInt16) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes a UInt32 in little-endian byte order.
    public mutating func writeU32(_ value: UInt32) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes a UInt64 in little-endian byte order.
    public mutating func writeU64(_ value: UInt64) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes an Int16 in little-endian byte order.
    public mutating func writeI16(_ value: Int16) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes an Int32 in little-endian byte order.
    public mutating func writeI32(_ value: Int32) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes an Int64 in little-endian byte order.
    public mutating func writeI64(_ value: Int64) {
        withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) }
    }

    /// Writes a Float (IEEE 754 binary32) in little-endian byte order.
    public mutating func writeF32(_ value: Float) {
        writeU32(value.bitPattern)
    }

    /// Writes a Double (IEEE 754 binary64) in little-endian byte order.
    public mutating func writeF64(_ value: Double) {
        writeU64(value.bitPattern)
    }
}
